sklearn/manifold/t_sne.py

```diff
@@ -753,34 +756,40 @@ def _fit(self, X, skip_num_points=0):
         if self.method == 'barnes_hut':
             if self.verbose:
                 print("[t-SNE] Computing %i nearest neighbors..." % k)
+
+            # Find the nearest neighbors for every point
+            # TODO: argument for class knn_estimator=None
+            # TODO: assert that the knn metric is euclidean
             if self.metric == 'precomputed':
-                # Use the precomputed distances to find
-                # the k nearest neighbors and their distances
-                neighbors_nn = np.argsort(distances, axis=1)[:, :k]
+                knn = NearestNeighbors(metric=self.metric)
             else:
-                # Find the nearest neighbors for every point
-                bt = BallTree(X)
-                # LvdM uses 3 * perplexity as the number of neighbors
-                # And we add one to not count the data point itself
-                # In the event that we have very small # of points
-                # set the neighbors to n - 1
-                distances_nn, neighbors_nn = bt.query(X, k=k + 1)
-                neighbors_nn = neighbors_nn[:, 1:]
-            P = _joint_probabilities_nn(distances, neighbors_nn,
+                knn = NearestNeighbors(algorithm='ball_tree',
+                                       metric=self.metric)
+            knn.fit(X)
+            # LvdM uses 3 * perplexity as the number of neighbors
+            # And we add one to not count the data point itself
+            # In the event that we have very small # of points
+            # set the neighbors to n - 1
+            distances_nn, neighbors_nn = knn.kneighbors(
```

Comment:
we could use `kneighbors_graph` which will already return a sparse matrix representation. This will allow us to rely on `kneighbors_graph` being as memory efficient as possible, rather than putting it into a sparse matrix in `_joint_probabilities_nn`.